File: app/src/main/java/org/ncssar/rid2caltopo/data/AircraftReadiness.kt
package org.ncssar.rid2caltopo.data

import org.json.JSONArray
import org.json.JSONObject
import java.io.Serializable

data class AircraftAccessory(
    val id: String,
    val name: String,
    val weightGrams: Double? = null,
    val required: Boolean = false,
    val group: String = ""
) : Serializable

data class AircraftReadiness(
    val serialNumber: String = "",
    val registrationNumber: String = "",
    val baseWeightGrams: Double? = null,
    val baseWeightIncludes: String = "",
    val requiredEquipment: String = "",
    val monitoringEquipment: String = "",
    val accessories: List<AircraftAccessory> = emptyList(),
    val recordId: String = ""
) : Serializable {
    fun validationErrors(): List<String> = buildList {
        if (baseWeightGrams != null && (!baseWeightGrams.isFinite() || baseWeightGrams !in 0.0..100000.0)) add("Invalid base weight.")
        if (accessories.size > 32 || accessories.map { it.id }.toSet().size != accessories.size) add("Accessory identifiers must be unique (maximum 32).")
        accessories.forEach {
            if (it.name.isBlank()) add("Give each accessory or battery a name.")
            if (it.weightGrams != null && (!it.weightGrams.isFinite() || it.weightGrams !in 0.0..100000.0)) add("Invalid accessory weight.")
        }
    }

    fun totalWeight(selected: Set<String>, payloadDescription: String, payloadGrams: Double?): Double? {
        if (selected.any { id -> accessories.none { it.id == id } }) return null
        val installed = accessories.filter { it.id in selected }
        if (installed.filter { it.group.isNotBlank() }.groupBy { it.group }.any { it.value.size > 1 }) return null
        if (baseWeightGrams == null || installed.any { it.weightGrams == null } ||
            (payloadDescription.isNotBlank() && payloadGrams == null)) return null
        val weights = listOf(baseWeightGrams, payloadGrams ?: 0.0) + installed.map { it.weightGrams!! }
        if (weights.any { !it.isFinite() || it < 0 }) return null
        return weights.sum()
    }

    fun toJSON(): JSONObject = JSONObject().apply {
        put("recordId", recordId)
        put("serialNumber", serialNumber); put("registrationNumber", registrationNumber)
        put("baseWeightGrams", baseWeightGrams ?: JSONObject.NULL)
        put("baseWeightIncludes", baseWeightIncludes); put("requiredEquipment", requiredEquipment)
        put("monitoringEquipment", monitoringEquipment)
        put("accessories", JSONArray().apply { accessories.forEach { accessory ->
            put(JSONObject().apply {
                put("id", accessory.id); put("name", accessory.name)
                put("weightGrams", accessory.weightGrams ?: JSONObject.NULL)
                put("required", accessory.required); put("group", accessory.group)
            })
        } })
    }

    companion object {
        @JvmStatic fun fromJSON(value: JSONObject?): AircraftReadiness {
            if (value == null) return AircraftReadiness()
            val items = value.optJSONArray("accessories") ?: JSONArray()
            return AircraftReadiness(
                value.optString("serialNumber"), value.optString("registrationNumber"),
                value.optionalWeight("baseWeightGrams"), value.optString("baseWeightIncludes"),
                value.optString("requiredEquipment"), value.optString("monitoringEquipment"),
                (0 until items.length()).map { index -> items.getJSONObject(index).let {
                    AircraftAccessory(it.getString("id"), it.getString("name"),
                        it.optionalWeight("weightGrams"), it.optBoolean("required"), it.optString("group"))
                } }, value.optString("recordId")
            )
        }
        private fun JSONObject.optionalWeight(key: String): Double? =
            if (!has(key) || isNull(key)) null else getDouble(key).also {
                require(it.isFinite() && it >= 0 && it <= 100000) { "Invalid weight in grams" }
            }
    }
}

File: apple/Sources/R2CCore/AircraftReadiness.swift
import Foundation

public struct AircraftAccessory: Codable, Sendable, Equatable, Identifiable {
    public var id: String
    public var name: String
    public var weightGrams: Double?
    public var required: Bool
    public var group: String

    public init(id: String = UUID().uuidString, name: String = "", weightGrams: Double? = nil,
                required: Bool = false, group: String = "") {
        self.id = id; self.name = name; self.weightGrams = weightGrams
        self.required = required; self.group = group
    }
}

public struct AircraftReadiness: Codable, Sendable, Equatable {
    public var recordId = ""
    public var serialNumber = ""
    public var registrationNumber = ""
    public var baseWeightGrams: Double?
    public var baseWeightIncludes = ""
    public var requiredEquipment = ""
    public var monitoringEquipment = ""
    public var accessories: [AircraftAccessory] = []

    public init() {}

    public func totalWeight(selected: Set<String>, payloadDescription: String, payloadGrams: Double?) -> Double? {
        guard selected.isSubset(of: Set(accessories.map(\.id))) else { return nil }
        let installed = accessories.filter { selected.contains($0.id) }
        let groups = Dictionary(grouping: installed.filter { !$0.group.isEmpty }, by: \.group)
        guard !groups.values.contains(where: { $0.count > 1 }),
              let baseWeightGrams, installed.allSatisfy({ $0.weightGrams != nil }),
              payloadDescription.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty || payloadGrams != nil
        else { return nil }
        let weights = [baseWeightGrams, payloadGrams ?? 0] + installed.compactMap(\.weightGrams)
        guard weights.allSatisfy({ $0.isFinite && $0 >= 0 }) else { return nil }
        return weights.reduce(0, +)
    }

    public var jsonString: String {
        guard let data = try? JSONEncoder().encode(self) else { return "{}" }
        return String(decoding: data, as: UTF8.self)
    }

    public var dictionary: [String: Any] {
        (try? JSONSerialization.jsonObject(with: Data(jsonString.utf8))) as? [String: Any] ?? [:]
    }

    public static func decode(_ value: Any?) -> AircraftReadiness {
        let data: Data?
        if let text = value as? String { data = text.data(using: .utf8) }
        else if let object = value as? [String: Any] { data = try? JSONSerialization.data(withJSONObject: object) }
        else { data = nil }
        return data.flatMap { try? JSONDecoder().decode(Self.self, from: $0) } ?? Self()
    }
}
